### PR TITLE
feat: swa build 🔨

### PR DESCRIPTION
**Pull request by sinedied**

### CODE REVIEW

File: e2e/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+node_modules/
+dist/
+package-lock.json
+swa-cli.config.json
```
Comment (manekinekko) [inline e2e/.gitignore:4]: is this a typo?

File: src/cli/commands/build.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    options: {
+      appLocation: options.appLocation,
+      apiLocation: options.apiLocation,
+      outputLocation: options.outputLocation,
+      appBuildCommand: options.appBuildCommand,
+      apiBuildCommand: options.apiBuildCommand,
+    }
+  });
+
+  let appLocation = options.appLocation !== undefined ? options.appLocation : workflowConfig?.appLocation;
```
Comment (manekinekko) [inline src/cli/commands/build.ts:42]:
```suggestion
  let appLocation = options.appLocation ?? workflowConfig?.appLocation;
```

File: src/cli/commands/build.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      logger.error(`Cannot generate your build configuration:`);
+      logger.error(error as Error, true);
+      return;
+    }
+  }
+  
+  if (!appBuildCommand && !apiBuildCommand) {
+    if (!hasBuildOptionsDefined(options)) {
+      logger.warn('No build options were defined.');
+      logger.warn(`Please "swa init" to set your configuration or use options flags to set your`);
```
Comment (manekinekko) [inline src/cli/commands/build.ts:84]:
```suggestion
      logger.warn(`Please run "swa init" to set your project configuration or use option flags to set your`);
```

File: src/cli/commands/build.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  logger.log(`- App location: ${chalk.green(appLocation || '')}`);
+  logger.log(`- API location: ${chalk.green(apiLocation || '')}`);
+  logger.log(`- Output location: ${chalk.green(outputLocation || '')}`);
+  logger.log(`- App build command: ${chalk.green(appBuildCommand || '')}`);
+  logger.log(`- API build command: ${chalk.green(apiBuildCommand || '')}`);
+  
+  if (appBuildCommand) {
+    let buildPath = appLocation!;
+    const packageJsonPath = await findUpPackageJsonDir(appLocation!, outputLocation!);
+    if (packageJsonPath) {
```
Comment (manekinekko) [inline src/cli/commands/build.ts:102]: what about python and dotnet?

File: src/cli/commands/build.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+function hasBuildOptionsDefined(options: SWACLIConfig): boolean {
+  if (options.appBuildCommand || options.apiBuildCommand) {
+    return true;
+  }
+  return isUserOrConfigOption('appBuildCommand') || isUserOrConfigOption('apiBuildCommand');
+}
+
+function showAutoErrorMessageAndExit() {
+  logger.error(`Please "swa init" to set your configuration or use options flags to set your`);
```
Comment (manekinekko) [inline src/cli/commands/build.ts:134]:
```suggestion
  logger.error(`Please run "swa init" to set your configuration or use option flags to set your`);
```

File: src/cli/commands/build.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  logger.error(`build commands and paths.`, true);
+}
+
+declare type NpmPackageManager = 'npm' | 'yarn' | 'pnpm';
```
Comment (manekinekko) [inline src/cli/commands/build.ts:138]: move to types.

File: src/cli/commands/build.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    logger.log(`Installing dependencies with "${packageManager} install"...`);
+    runCommand(`${packageManager} install`, packageJsonPath);
+  } catch (error) {
+    logger.error(`Cannot install dependencies:`);
+    logger.error(error as Error, true);
+  }
+}
+
+function runCommand(command: string, cwd: string) {
+  execSync(command, {
```
Comment (manekinekko) [inline src/cli/commands/build.ts:168]: we need to make sure the command here is windows safe. We had issues in the past with commands not executing on Windows.

**Review by manekinekko — CHANGES_REQUESTED** [review]

File: e2e/.gitignore (hunk 1 above)
Comment (sinedied) [reply e2e/.gitignore:4]: No, this is to keep the e2e/samples and e2e/fixtures folders clean.
Or if there's a typo inside, I don't see it 😆

File: src/cli/commands/start.ts [hunk 8]
```diff
@@ -307,17 +306,30 @@ export async function start(outputLocationOtHttpUrl: string, options: SWACLIConf
...
     );
   }
 
-  if (build) {
-    // run the app/api builds
-    await builder({
-      config: userWorkflowConfig as GithubActionWorkflow,
+  if (runBuild) {
+    // run swa build
+    execSync("npx swa build", {
```
[the remark this answers is not in the document]
Comment (sinedied) [reply src/cli/commands/start.ts:330]: I'll remove npx as the new process inherits current process.env from NPM, npx is not needed here (forgot to remove it)

File: src/swa.d.ts [hunk 9]
```diff
@@ -114,7 +114,7 @@ declare type GithubActionWorkflow = {
 declare type SWACLIOptionsToCleanUp = {
   // TODO: cleanup
   // app?: string;
-  build?: boolean;
+  // build?: boolean;
```
[the remark this answers is not in the document]
Comment (sinedied) [reply src/swa.d.ts:117]: I cleaned these in the PR that will follow :)

File: src/cli/commands/build.ts (hunk 4 above)
Comment (sinedied) [reply src/cli/commands/build.ts:102]: They will run from appLocation by default, but this is not ideal.
We should open a separate issue to address how we find the root app and api paths for all stacks, to use that with "swa build" and "swa start" (as it's also needed for the --run flag)

File: src/cli/commands/build.ts (hunk 6 above)
Comment (sinedied) [reply src/cli/commands/build.ts:138]: whoops, forgot that one, good catch!

File: src/cli/commands/build.ts (hunk 7 above)
Comment (sinedied) [reply src/cli/commands/build.ts:168]: I have a few other tools relying on the same code that are working fine on windows for years ;)
We should be able to test it in Cypress anyways

File: src/cli/commands/start.ts [hunk 10]
```diff
@@ -307,17 +306,30 @@ export async function start(outputLocationOtHttpUrl: string, options: SWACLIConf
...
   if (startupCommand) {
+    let startupPath = userWorkflowConfig?.appLocation;
+    const packageJsonDir = await findUpPackageJsonDir(
+      userWorkflowConfig?.appLocation || '.',
+      // TODO: incorrect when using devServer, need refacto of context
+      userWorkflowConfig?.outputLocation || '.'
+    );
+    if (packageJsonDir) {
+      logger.silly(`Found package.json in ${packageJsonDir}`);
+      // if a package.json file is found, use its directory as the startup path
```
[the remark this answers is not in the document]
Comment (sinedied) [reply src/cli/commands/start.ts:319]: same as my previous command about dotnet/python

**Pull request merged**